File: packages/google_mlkit_entity_extraction/ios/Classes/GoogleMlKitEntityExtractionPlugin.m
#import "GoogleMlKitEntityExtractionPlugin.h"
#import <MLKitEntityExtraction/MLKitEntityExtraction.h>
#import <google_mlkit_commons/GoogleMlKitCommonsPlugin.h>

#define channelName @"google_mlkit_entity_extractor"
#define startEntityExtractor @"nlp#startEntityExtractor"
#define closeEntityExtractor @"nlp#closeEntityExtractor"
#define manageEntityExtractionModels @"nlp#manageEntityExtractionModels"

@implementation GoogleMlKitEntityExtractionPlugin {
    NSMutableDictionary *instances;
    GenericModelManager *genericModelManager;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:channelName
                                     binaryMessenger:[registrar messenger]];
    GoogleMlKitEntityExtractionPlugin* instance = [[GoogleMlKitEntityExtractionPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (id)init {
    self = [super init];
    if (self)
        instances = [NSMutableDictionary dictionary];
    return  self;
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([call.method isEqualToString:startEntityExtractor]) {
        [self handleDetection:call result:result];
    } else if ([call.method isEqualToString:manageEntityExtractionModels]) {
        [self manageModel:call result:result];
    } else if ([call.method isEqualToString:closeEntityExtractor]) {
        NSString *uid = call.arguments[@"id"];
        [instances removeObjectForKey:uid];
        result(NULL);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)handleDetection:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSString *text = call.arguments[@"text"];
    
    NSString *uid = call.arguments[@"id"];
    MLKEntityExtractor *entityExtractor = [instances objectForKey:uid];
    if (entityExtractor == NULL) {
        NSString *language = call.arguments[@"language"];
        MLKEntityExtractorOptions *options = [[MLKEntityExtractorOptions alloc] initWithModelIdentifier:language];
        entityExtractor = [MLKEntityExtractor entityExtractorWithOptions:options];
        instances[uid] = entityExtractor;
    }
    
    MLKEntityExtractionParams *params = [[MLKEntityExtractionParams alloc] init];
    NSDictionary *parameters = call.arguments[@"parameters"];
    
    NSString *timezone = parameters[@"timezone"];
    if ([timezone isKindOfClass: [NSString class]] && timezone.length > 0) {
        params.referenceTimeZone = [NSTimeZone timeZoneWithAbbreviation:timezone];
    }

    NSString *time = parameters[@"time"];
    if ([time isKindOfClass: [NSNumber class]]) {
        // NSTimeInterval should is expressed in seconds, not milliseconds
        params.referenceTime = [NSDate dateWithTimeIntervalSince1970: time.doubleValue / 1000];
    }
    
    NSString *locale = parameters[@"locale"];
    if ([locale isKindOfClass: [NSString class]] && locale.length > 0) {
        params.preferredLocale = [NSLocale localeWithLocaleIdentifier:locale];
    }
    
    NSArray *filtersValues = parameters[@"filters"];
    if ([filtersValues isKindOfClass: [NSArray class]] && filtersValues.count > 0) {
        NSMutableSet *filters = [NSMutableSet set];
        for(NSNumber *number in filtersValues) {
            int value = number.intValue;
            switch(value) {
                case 1:
                    [filters addObject:MLKEntityExtractionEntityTypeAddress];
                    break;
                case 2:
                    [filters addObject:MLKEntityExtractionEntityTypeDateTime];
                    break;
                case 3:
                    [filters addObject:MLKEntityExtractionEntityTypeEmail];
                    break;
                case 4:
                    [filters addObject:MLKEntityExtractionEntityTypeFlightNumber];
                    break;
                case 5:
                    [filters addObject:MLKEntityExtractionEntityTypeIBAN];
                    break;
                case 6:
                    [filters addObject:MLKEntityExtractionEntityTypeISBN];
                    break;
                case 7:
                    [filters addObject:MLKEntityExtractionEntityTypePaymentCard];
                    break;
                case 8:
                    [filters addObject:MLKEntityExtractionEntityTypePhone];
                    break;
                case 9:
                    [filters addObject:MLKEntityExtractionEntityTypeTrackingNumber];
                    break;
                case 10:
                    [filters addObject:MLKEntityExtractionEntityTypeURL];
                    break;
                case 11:
                    [filters addObject:MLKEntityExtractionEntityTypeMoney];
                    break;
                default:
                    break;
            }
        }
        params.typesFilter = filters;
    }
    
    [entityExtractor downloadModelIfNeededWithCompletion:^(NSError *_Nullable error) {
        if (error) {
            result(getFlutterError(error));
            return;
        }
        // Model downloaded successfully. Okay to annotate.
        
        [entityExtractor annotateText:text
                           withParams:params
                           completion:^(NSArray *_Nullable annotations,
                                        NSError *_Nullable error) {
            if (error) {
                result(getFlutterError(error));
                return;
            } else if (!annotations) {
                result(NULL);
                return;
            }
            
            NSMutableArray *allAnnotations = [NSMutableArray array];
            for (MLKEntityAnnotation *annotation in annotations) {
                NSMutableDictionary *data = [NSMutableDictionary dictionary];
                data[@"text"] = [text substringWithRange:annotation.range];
                data[@"start"] = @((int)annotation.range.location);
                data[@"end"] = @((int)(annotation.range.location + annotation.range.length));
                
                NSMutableArray *allEntities = [NSMutableArray array];
                NSArray *entities = annotation.entities;
                for (MLKEntity *entity in entities) {
                    NSMutableDictionary *entityData = [NSMutableDictionary dictionary];
                    int type = 0;
                    
                    if ([entity.entityType isEqualToString: MLKEntityExtractionEntityTypeAddress]) {
                        type = 1;
                    } else if ([entity.entityType isEqualToString: MLKEntityExtractionEntityTypeDateTime]) {
                        type = 2;
                        entityData[@"dateTimeGranularity"] = @(entity.dateTimeEntity.dateTimeGranularity);
                        // result is expected in milliseconds, not seconds.
                        entityData[@"timestamp"] = @(entity.dateTimeEntity.dateTime.timeIntervalSince1970 * 1000);
                    } else if ([entity.entityType isEqualToString: MLKEntityExtractionEntityTypeEmail]) {
                        type = 3;
                    } else if ([entity.entityType isEqualToString: MLKEntityExtractionEntityTypeFlightNumber]) {
                        type = 4;
                        entityData[@"code"] = entity.flightNumberEntity.airlineCode;
                        entityData[@"number"] = entity.flightNumberEntity.flightNumber;
                    } else if ([entity.entityType isEqualToString: MLKEntityExtractionEntityTypeIBAN]) {
                        type = 5;
                        entityData[@"iban"] = entity.IBANEntity.IBAN;
                        entityData[@"code"] = entity.IBANEntity.countryCode;
                    } else if ([entity.entityType isEqualToString: MLKEntityExtractionEntityTypeISBN]) {
                        type = 6;
                        entityData[@"isbn"] = entity.ISBNEntity.ISBN;
                    } else if ([entity.entityType isEqualToString: MLKEntityExtractionEntityTypePaymentCard]) {
                        type = 7;
                        entityData[@"network"] = @(entity.paymentCardEntity.paymentCardNetwork);
                        entityData[@"number"] = entity.paymentCardEntity.paymentCardNumber;
                    } else if ([entity.entityType isEqualToString: MLKEntityExtractionEntityTypePhone]) {
                        type = 8;
                    } else if ([entity.entityType isEqualToString: MLKEntityExtractionEntityTypeTrackingNumber]) {
                        type = 9;
                        entityData[@"carrier"] = @(entity.trackingNumberEntity.parcelCarrier);
                        entityData[@"number"] = entity.trackingNumberEntity.parcelTrackingNumber;
                    } else if ([entity.entityType isEqualToString: MLKEntityExtractionEntityTypeURL]) {
                        type = 10;
                    } else if ([entity.entityType isEqualToString: MLKEntityExtractionEntityTypeMoney]) {
                       type = 11;
                       entityData[@"fraction"] = @(entity.moneyEntity.fractionalPart);
                       entityData[@"integer"] = @(entity.moneyEntity.integerPart);
                       entityData[@"unnormalized"] = entity.moneyEntity.unnormalizedCurrency;
                    }
                    
                    entityData[@"type"] = @(type);
                    entityData[@"raw"] = [NSString stringWithFormat:@"%@", entity];
                    
                    [allEntities addObject:entityData];
                }
                data[@"entities"] = allEntities;
                [allAnnotations addObject:data];
            }
            
            result(allAnnotations);
        }];
    }];
}

- (void)manageModel:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSString *modelTag = call.arguments[@"model"];
    MLKEntityExtractionRemoteModel *model =  [MLKEntityExtractionRemoteModel entityExtractorRemoteModelWithIdentifier:modelTag];
    genericModelManager = [[GenericModelManager alloc] init];
    [genericModelManager manageModel:model call:call result:result];
}

@end

File: packages/google_mlkit_entity_extraction/ios/google_mlkit_entity_extraction/Sources/google_mlkit_entity_extraction/GoogleMlKitEntityExtractionPlugin.swift
import Flutter
import google_mlkit_commons

#if canImport(MLKitEntityExtraction)
    import MLKitEntityExtraction
#endif

@objc
public class GoogleMlKitEntityExtractionPlugin: NSObject, FlutterPlugin {
    private var instances: [String: Any] = [:]
    private var genericModelManager: GenericModelManager?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "google_mlkit_entity_extractor",
            binaryMessenger: registrar.messenger()
        )
        let instance = GoogleMlKitEntityExtractionPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        #if canImport(MLKitEntityExtraction)
            switch call.method {
            case "nlp#startEntityExtractor":
                handleDetection(call: call, result: result)
            case "nlp#manageEntityExtractionModels":
                manageModel(call: call, result: result)
            case "nlp#closeEntityExtractor":
                if let args = call.arguments as? [String: Any], let uid = args["id"] as? String {
                    instances.removeValue(forKey: uid)
                }
                result(nil)
            default:
                result(FlutterMethodNotImplemented)
            }
        #else
            let unimplemented = FlutterError(
                code: "UNIMPLEMENTED",
                message: "MLKitEntityExtraction is not available via Swift Package Manager on iOS",
                details: nil
            )
            switch call.method {
            case "nlp#closeEntityExtractor":
                if let args = call.arguments as? [String: Any], let uid = args["id"] as? String {
                    instances.removeValue(forKey: uid)
                }
                result(nil)
            default:
                result(unimplemented)
            }
        #endif
    }

    #if canImport(MLKitEntityExtraction)
        private func handleDetection(call: FlutterMethodCall, result: @escaping FlutterResult) {
            guard let args = call.arguments as? [String: Any],
                let text = args["text"] as? String,
                let uid = args["id"] as? String
            else {
                result(
                    FlutterError(code: "invalid_args", message: "Missing arguments", details: nil))
                return
            }

            let entityExtractor: EntityExtractor
            if let existing = instances[uid] as? EntityExtractor {
                entityExtractor = existing
            } else {
                guard let language = args["language"] as? String else {
                    result(
                        FlutterError(
                            code: "invalid_args", message: "Missing language", details: nil))
                    return
                }
                let modelIdentifier = EntityExtractionModelIdentifier(rawValue: language)
                let options = EntityExtractorOptions(modelIdentifier: modelIdentifier)
                entityExtractor = EntityExtractor.entityExtractor(options: options)
                instances[uid] = entityExtractor
            }

            let params = EntityExtractionParams()
            let parametersRaw = args["parameters"]
            let parameters: [String: Any] =
                (parametersRaw is NSNull) ? [:] : (parametersRaw as? [String: Any]) ?? [:]
            if let timezone = valueAsString(parameters["timezone"]), !timezone.isEmpty {
                params.referenceTimeZone = TimeZone(abbreviation: timezone)
            }
            if let time = valueAsNumber(parameters["time"]) {
                params.referenceTime = Date(timeIntervalSince1970: time.doubleValue / 1000)
            }
            if let locale = valueAsString(parameters["locale"]), !locale.isEmpty {
                params.preferredLocale = Locale(identifier: locale)
            }
            if let filtersValues = valueAsNumberArray(parameters["filters"]), !filtersValues.isEmpty {
                let filters: Set<EntityType> = Set(
                    filtersValues.compactMap { numberToEntityType($0.intValue) })
                params.typesFilter = filters
            }

            entityExtractor.downloadModelIfNeeded { error in
                if let error = error as NSError? {
                    result(
                        FlutterError(
                            code: "Error \(error.code)", message: error.domain,
                            details: error.localizedDescription))
                    return
                }
                entityExtractor.annotateText(text, params: params) { annotations, error in
                    if let error = error as NSError? {
                        result(
                            FlutterError(
                                code: "Error \(error.code)", message: error.domain,
                                details: error.localizedDescription))
                        return
                    }
                    guard let annotations = annotations else {
                        result(nil)
                        return
                    }
                    let allAnnotations = annotations.map { annotation -> [String: Any] in
                        let range = annotation.range
                        let substring = (text as NSString).substring(with: range)
                        let entities = annotation.entities.map { entity in
                            self.entityToDictionary(text: text, entity: entity)
                        }
                        return [
                            "text": substring,
                            "start": range.location,
                            "end": range.location + range.length,
                            "entities": entities
                        ]
                    }
                    result(allAnnotations)
                }
            }
        }

        private func numberToEntityType(_ value: Int) -> EntityType? {
            switch value {
            case 1: return .address
            case 2: return .dateTime
            case 3: return .email
            case 4: return .flightNumber
            case 5: return .IBAN
            case 6: return .ISBN
            case 7: return .paymentCard
            case 8: return .phone
            case 9: return .trackingNumber
            case 10: return .URL
            case 11: return .money
            default: return nil
            }
        }

        private func entityToDictionary(text: String, entity: Entity) -> [String: Any] {
            var entityData: [String: Any] = [
                "type": entityTypeToNumber(entity.entityType),
                "raw": String(describing: entity)
            ]
            if entity.entityType == .dateTime, let dateTimeEntity = entity.dateTimeEntity {
                entityData["dateTimeGranularity"] = dateTimeEntity.dateTimeGranularity.rawValue
                entityData["timestamp"] = dateTimeEntity.dateTime.timeIntervalSince1970 * 1000
            } else if entity.entityType == .flightNumber,
                let flightEntity = entity.flightNumberEntity {
                entityData["code"] = flightEntity.airlineCode
                entityData["number"] = flightEntity.flightNumber
            } else if entity.entityType == .IBAN, let ibanEntity = entity.ibanEntity {
                entityData["iban"] = ibanEntity.iban
                entityData["code"] = ibanEntity.countryCode
            } else if entity.entityType == .ISBN, let isbnEntity = entity.isbnEntity {
                entityData["isbn"] = isbnEntity.isbn
            } else if entity.entityType == .paymentCard, let cardEntity = entity.paymentCardEntity {
                entityData["network"] = cardEntity.paymentCardNetwork.rawValue
                entityData["number"] = cardEntity.paymentCardNumber
            } else if entity.entityType == .trackingNumber,
                let trackingEntity = entity.trackingNumberEntity {
                entityData["carrier"] = trackingEntity.parcelCarrier.rawValue
                entityData["number"] = trackingEntity.parcelTrackingNumber
            } else if entity.entityType == .money, let moneyEntity = entity.moneyEntity {
                entityData["fraction"] = moneyEntity.fractionalPart
                entityData["integer"] = moneyEntity.integerPart
                entityData["unnormalized"] = moneyEntity.unnormalizedCurrency
            }
            return entityData
        }

        private func entityTypeToNumber(_ type: EntityType) -> Int {
            switch type {
            case .address: return 1
            case .dateTime: return 2
            case .email: return 3
            case .flightNumber: return 4
            case .IBAN: return 5
            case .ISBN: return 6
            case .paymentCard: return 7
            case .phone: return 8
            case .trackingNumber: return 9
            case .URL: return 10
            case .money: return 11
            default: return 0
            }
        }

        private func valueAsString(_ value: Any?) -> String? {
            guard value != nil, value is NSNull == false else { return nil }
            return value as? String
        }

        private func valueAsNumber(_ value: Any?) -> NSNumber? {
            guard value != nil, value is NSNull == false else { return nil }
            return value as? NSNumber
        }

        private func valueAsNumberArray(_ value: Any?) -> [NSNumber]? {
            guard value != nil, value is NSNull == false else { return nil }
            return value as? [NSNumber]
        }

        private func manageModel(call: FlutterMethodCall, result: @escaping FlutterResult) {
            guard let args = call.arguments as? [String: Any],
                let modelTag = args["model"] as? String
            else {
                result(
                    FlutterError(
                        code: "invalid_args", message: "Missing model argument", details: nil))
                return
            }
            let modelIdentifier = EntityExtractionModelIdentifier(rawValue: modelTag)
            let model = EntityExtractorRemoteModel.entityExtractorRemoteModel(
                identifier: modelIdentifier)
            if genericModelManager == nil {
                genericModelManager = GenericModelManager()
            }
            genericModelManager?.manage(model: model, call: call, result: result)
        }
    #endif
}
